Sources/Diff/Diff.swift

```diff
@@ -44,70 +44,79 @@ private let plus = "+"
 
   let (hunk, hunks) = diffs
     .reduce((current: Hunk(), hunks: [Hunk]())) { cursor, diff in
-      var (current, hunks) = cursor
+      let (current, hunks) = cursor
       let len = diff.elements.count
 
       switch diff.type {
+      case .both where len > ctx * 2:
+        let hunk = current + Hunk(len: ctx, lines: diff.elements.prefix(ctx).map(prepending(figureSpace)))
+        let next = Hunk(
+          fstIdx: current.fstIdx + current.fstLen + len - ctx,
+          fstLen: ctx,
+          sndIdx: current.sndIdx + current.sndLen + len - ctx,
+          sndLen: ctx,
+          lines: (diff.elements.suffix(ctx) as ArraySlice<S>).map(prepending(figureSpace))
+        )
+        return (next, changed(hunk) ? hunks + [hunk] : hunks)
+      case .both where current.lines.isEmpty:
+        let lines = (diff.elements.suffix(ctx) as ArraySlice<S>).map(prepending(figureSpace))
+        let count = lines.count
+        return (current + Hunk(idx: len - count, len: count, lines: lines), hunks)
       case .both:
-        if len > ctx * 2 {
-          current.fstLen += ctx
-          current.sndLen += ctx
-          current.lines.append(contentsOf: diff.elements.prefix(ctx).map(prepending(figureSpace)))
-          if current.lines.contains(where: { $0.hasPrefix(minus) || $0.hasPrefix(plus) }) {
-            hunks.append(current)
-          }
-
-          current.fstIdx += len + 1
-          current.fstLen = ctx
-          current.sndIdx += len + 1
-          current.sndLen = ctx
-          current.lines = (diff.elements.suffix(ctx) as ArraySlice<S>).map(prepending(figureSpace))
-        } else if current.lines.isEmpty {
-          let lines = (diff.elements.suffix(ctx) as ArraySlice<S>).map(prepending(figureSpace))
-          let count = lines.count
-          current.fstIdx += len - count
-          current.fstLen += count
-          current.sndIdx += len - count
-          current.sndLen += count
-          current.lines.append(contentsOf: lines)
-        } else {
-          current.fstLen += len
-          current.sndLen += len
-          current.lines.append(contentsOf: diff.elements.map(prepending(figureSpace)))
-        }
-        return (current, hunks)
+        return (current + Hunk(len: len, lines: diff.elements.map(prepending(figureSpace))), hunks)
       case .first:
-        current.fstLen += len
-        current.lines.append(contentsOf: diff.elements.map(prepending(minus)))
+        return (current + Hunk(fstLen: len, lines: diff.elements.map(prepending(minus))), hunks)
       case .second:
-        current.sndLen += len
-        current.lines.append(contentsOf: diff.elements.map(prepending(plus)))
+        return (current + Hunk(sndLen: len, lines: diff.elements.map(prepending(plus))), hunks)
       }
-
-      return (current, hunks)
   }
 
-  return hunk.lines.contains(where: { $0.hasPrefix(minus) || $0.hasPrefix(plus) })
-    ? hunks + [hunk]
-    : hunks
+  return changed(hunk) ? hunks + [hunk] : hunks
```

Comment:
Feel like there's a logical way to change the various `where` logic in the switch above to avoid this repetition, but haven't dug too deep.